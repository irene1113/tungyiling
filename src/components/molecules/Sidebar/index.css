.sidebar {
	display: inline-flex;
	position: fixed;
	height: 100vh;
	width: 0;
	z-index: 200;
	transition: width 0.3s;

	& > .hamburger {
		position: fixed;
		left: -60px;
		z-index: 20;
		transform: translateY(-50%);
		transition: left 0.3s;
		cursor: pointer;

		&.is-show {
			left: 0;
		}
	}

	& > .arc {
		display: inline-flex;
		flex-shrink: 0;
		background: var(--primary);
		border-radius: 50%;
		height: 600vh;
		width: 600vh;
		position: fixed;
		right: 0%;
		top: 0;
		transform: translateY(-50%);
		transition: right 0.3s;

		&.is-show {
			right: -60px;
		}
	}

	& > .nav {
		display: flex;
		flex-direction: column;
		padding-top: 60px;
		padding-left: 60px;
		z-index: 10;

		& > .link-item {
			display: flex;

			& > a {
				color: white;
				font-size: 8vw;
				font-weight: bold;
				margin: 10px 0;
				font-family: 'Asap', sans-serif;
			}
		}
	}

	&.open {
		width: 100vw;

		& > .hamburger {
			left: 0;
		}

		& > .arc {
			right: 10vw;
		}
	}

	& > .filter-svg {
		display: none;
		position: absolute;
	}
}
