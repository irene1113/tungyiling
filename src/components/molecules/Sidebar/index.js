/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect } from 'react';
import classnames from 'classnames';

import { useMousemove, useScroll } from 'util/event';
import { useBoolean } from 'util/hook';
import { useRouting } from 'models/routing';

import Link from 'components/atoms/Link';
import HamburgerIcon from 'images/icon/hamburger.inline.svg';
import CloseIcon from 'images/icon/close.inline.svg';

import styles from './index.css';

const ToggleIcon = ({ isOpen, openMenu, closeMenu, ...props }) => {
	return isOpen ? (
		<CloseIcon onClick={closeMenu} {...props} />
	) : (
		<HamburgerIcon onClick={openMenu} {...props} />
	);
};

const LinkItem = ({ className, to, active, closeMenu = () => {}, children }) => (
	<div className={classnames(styles.linkItem, className, active && styles.active)}>
		<Link to={to} className={classnames(className, active && styles.active)} onClick={closeMenu}>
			{children}
		</Link>
	</div>
);

const Sidebar = ({ className }) => {
	const [{ pathname }] = useRouting();
	const mousePosition = useMousemove();
	const scrollY = useScroll();
	const [isOpen, { setTrue: openMenu, setFalse: closeMenu }] = useBoolean();
	const [isShow, { setTrue: showMenu, setFalse: hideMenu }] = useBoolean({ defaultBoolean: true });

	useEffect(() => {
		if (scrollY <= 20) {
			showMenu();
		} else if (mousePosition.x <= 100 && !isShow) {
			showMenu();
		} else if (mousePosition.x > 100 && isShow) {
			hideMenu();
		}
	}, [mousePosition, scrollY]);

	return (
		<div
			className={classnames(styles.sidebar, className, { [styles.open]: isOpen })}
			style={{ filter: `url(#sticky-menu)` }}
		>
			<ToggleIcon
				className={classnames(styles.hamburger, { [styles.isShow]: isShow })}
				style={{
					top: `${isOpen ? 30 : mousePosition.y}px`,
				}}
				isOpen={isOpen}
				openMenu={openMenu}
				closeMenu={closeMenu}
			/>
			{isOpen && (
				<div className={styles.nav}>
					<LinkItem to="/projects" active={pathname === '/projects'} closeMenu={closeMenu}>
						PROJECTS
					</LinkItem>
					<LinkItem to="/about" active={pathname === '/about'} closeMenu={closeMenu}>
						ABOUT
					</LinkItem>
					<LinkItem to="/contact" active={pathname === '/contact'} closeMenu={closeMenu}>
						CONTACT
					</LinkItem>
				</div>
			)}

			<div
				className={classnames(styles.arc, { [styles.isShow]: isShow })}
				style={{
					top: `${mousePosition.y}px`,
				}}
			/>
		</div>
	);
};

export default Sidebar;
