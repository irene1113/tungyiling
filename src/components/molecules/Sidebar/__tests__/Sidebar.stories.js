import React from 'react';
import { storiesOf } from '@storybook/react';

import Sidebar from 'components/molecules/Sidebar';

const stories = storiesOf('molecules/Sidebar', module);

stories.add('__interactive', () => <Sidebar />, {
	redux: {
		data: {
			routing: {
				pathname: '',
			},
		},
	},
});
