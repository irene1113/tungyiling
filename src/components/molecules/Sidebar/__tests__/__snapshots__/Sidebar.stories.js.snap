// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="sidebar"
    style="filter: url(#sticky-menu);"
  >
    <test-file-stub
      classname="hamburger isShow"
      style="top: 100px;"
    />
    <div
      class="arc isShow"
      style="top: 100px;"
    />
  </div>
</DocumentFragment>
`;
